app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt

```diff
@@ -31,32 +29,69 @@ import kotlinx.coroutines.withContext
 class UserProfileViewModel @Inject constructor(
     private val navigationManager: NavigationManager,
     private val dataStore: UserDataStore,
-    private val uploadUserAvatar: UploadUserAvatarUseCase,
+    private val getPublicAssetUseCase: GetPublicAssetUseCase,
     private val getSelf: GetSelfUserUseCase
 ) : ViewModel() {
 
     var userProfileState by mutableStateOf(SelfUserProfileState())
         private set
 
     init {
-        // TODO: here we should have a loading state as the first initial state of the screen
         viewModelScope.launch {
-            getSelf().collect {
-                userProfileState = SelfUserProfileState(
-                    status = UserStatus.AVAILABLE,
-                    fullName = it.name.orEmpty(),
-                    userName = it.handle.orEmpty(),
-                    teamName = it.team,
-                    // Add some mocked team
-                    otherAccounts = listOf(
-                        OtherAccount("someId", "", it.name.orEmpty(), "Wire Swiss GmbH"),
-                        OtherAccount("someId", "", "B. A. Baracus", "The A-Team"),
+            fetchSelfUser()
+        }
+    }
+
+    private suspend fun fetchSelfUser() {
+        // Show progress bar on initial load
+        showLoadingAvatar(true)
+
+        viewModelScope.launch {
+            getSelf().collect { selfUser ->
+                with(selfUser) {
+                    // Load user avatar raw image data
+                    val avatarAssetRaw = completePicture?.let { getAvatarAsByteArrayOrNull(it) }
+
+                    // Update avatar asset id on user data store
+                    completePicture?.apply { dataStore.updateUserAvatarAssetId(this) }
+
+                    // Hide progress bar
+                    showLoadingAvatar(false)
+
+                    // Update user data state
+                    userProfileState = SelfUserProfileState(
+                        avatarAssetByteArray = avatarAssetRaw,
+                        status = UserStatus.AVAILABLE,
+                        fullName = name.orEmpty(),
+                        userName = handle.orEmpty(),
+                        teamName = team,
+
+                        // TODO: remove mocked team when other accounts functionality is ready
+                        otherAccounts = listOf(
+                            OtherAccount("someId", "", name.orEmpty(), "Wire Swiss GmbH"),
+                            OtherAccount("someId", "", "B. A. Baracus", "The A-Team"),
+                        )
```

Comment:
In my opinion a better user experience would be if we load the avatar seperetly as it may take more time because we are calling additional usecase for that, we could display user info immediatly though, what do you think ?

Reply:
That is indeed a good point. However I think 99% of the cases it will take more or less the same time to load both, as avatar is meant to change rarely and therefore the subsequent calls to this use case will be rather immediate as the avatar will be pulled directly from the assets database. I could still change the logic a bit, to load the asset on a separate coroutine, if you think it's indeed necessary and the current avatar loading times are too high though.

Reply:
I do not think it can hurt and may improve a user experience abit, but your call if you think it is worth it and will not take much time

Reply:
Good catch. Added